### DISCUSSION TITLE
Roadmap 2023

### DISCUSSION THREAD

**Discussion by pluralia** [eclipse-langium/langium#863] [Announcements]
Over the course of this year, the Langium contributors plan to work on the following topics:
- [ ] @msujew @pluralia Enhancing grammar rules after adding LL(*) in Chevrotain
- [x] @sailingKieler Generator tracing
- [ ] @msujew @Lotes Model serialization (stringification)
- [x] @pluralia @msujew @sailingKieler Type system support
- [x] @claaspetersen @emilkrebs Showcase for the web-site
- [ ] @montymxb @pluralia @Lotes Stabilizing type inference using formal methods

**Comment by trusktr** [comment]
> * [ ]  @pluralia @msujew @sailingKieler Type system support

This sounds really interesting!

- Any thought already on how it may look? How will it make defining a type system for a language easier? What might the API look like?
- What would be limitations of using this system as opposed to defining a language's type checking in a manual way like in langium-lox?

**Reply by msujew** [reply]
> What would be limitations of using this system as opposed to defining a language's type checking in a manual way like in langium-lox?

Let me answer that first: Ideally there'd be none. This is at least one of the design goals. In the "worst case" developers should be able to modify every little detail of how their type system behaves.

> Any thought already on how it may look? How will it make defining a type system for a language easier? What might the API look like?

I'm currently in the process of designing a rough outline of the type system API. There are basically 5 parts to how we usually implement type systems "manually" and we hope to eliminate 3-4 parts of that.

1. Type inference: I.e. the process of creating types from AST nodes. We will provide some utility to help with some very specific problems in this (such as inference of generic type arguments based on hints/argument types), but this is still highly linked to the semantics of a language. Therefore this will need a lot of help from the developer. The library is also supposed to help with type caching (this also applies to point 3).
2. Type definitions: Most languages use well-known types. Primitives, classes, interfaces, arrays, delegates, unions, intersections, top/bottom types (`any` and `never` in TypeScript). These should be contained as components in the library, which eliminate the need to implement them yourself. Additionally, developers should be able to define their own custom types.
3. Type assignability: For well known types, this should come straight out-of-the-box. Custom types should also somehow be included in here. To improve performance, caching is fairly important in here as well.
4. Operator support: Contains information on which types are compatible for which operators and what these operations return.
5. Type stringification: This is a trivial topic, but still important for a lot of LSP services, such as hover, signature help and completion.

**Reply by MechMel** [reply]
This is super cool. Also means a good chunk of my work the past week will be obsolete, but I'm okay with this.

**Reply by trusktr** [reply]
- How will this manifest itself in the Langium API that language makers use? For example, do language makers write type relationships inside the Grammar of their language? Will the AST that is output by a generated parser now include type information?
- Are there any particular type theories that will be used?

**Reply by spoenemann** [reply]
> How will this manifest itself in the Langium API that language makers use? For example, do language makers write type relationships inside the Grammar of their language? Will the AST that is output by a generated parser now include type information?

The type system implementation will likely be independent of Langium (might be even in a separate repo). As a language developer, it will be your responsibility to create adapters between AST and type system construct, but that will be easy to do.

**Comment by spoenemann** [comment]
Update for the roadmap: We plan to release v2.0.0 in early August. There will be a "release candidate" of some form that everybody can use to verify the new version in their projects and prepare necessary changes.

Though many of the broader topics mentioned by Irina here are not covered yet, the list of improvements is huge:
https://github.com/langium/langium/milestone/7?closed=1

**Comment by spoenemann** [comment]
Another update: it turned out that we didn't have time for many of the roadmap items this year. We'll go into planning what we can achieve in 2024.

**Reply by trusktr** [reply]
@spoenemann How did 2024 go, and what's the current vision for 2025?

I'm particularly still interested in Type Systems. All of my focus lately has been on getting an MBA and starting a business, but once I have a business rolling, I would like to put some resources towards continuing to build the language/compiler I'd started to make (TS to Wasm), so for that I'm keen to see how any type system APIs could help. P.S. would love to share the repo with you if you directly if you message me, and would love any feedback on our baby steps.

**Reply by msujew** [reply]
We've released our first public version of our type system library last month, see [typir](https://github.com/TypeFox/typir).

**Reply by trusktr** [reply]
Oh wow this is exciting!
